### ISSUE TITLE
Build user is not exists on macOS High Sierra

### ISSUE THREAD

**Issue by david50407** [NixOS/nix#1720]
I was on single-user mode for avoiding #1583.
And I upgrade my High Sierra to 10.13.2, so I just created Nix build group/users back following the installation script [here](https://github.com/NixOS/nix/blob/master/scripts/install-darwin-multi-user.sh).

But after that, when I want to update the `nixpkgs`, it tells my that build user is not exist, logs below:

```
$ nix-channel --update nixpkgs
downloading Nix expressions from ‘https://d3g5gsiof5omrk.cloudfront.net/nixpkgs/nixpkgs-18.03pre122127.79081322ab9/nixexprs.tar.xz’...
downloading ‘https://d3g5gsiof5omrk.cloudfront.net/nixpkgs/nixpkgs-18.03pre122127.79081322ab9/nixexprs.tar.xz’... [7439/9432 KiB, 741.9 KiB/s]
unpacking channels...
error: the user ‘nixbld1’ in the group ‘nixbld’ does not exist
cannot unpack the channels at /nix/var/nix/profiles/default/bin/nix-channel line 172.
```

Build group informations:

```
$ dscl . -read /Groups/nixbld
dsAttrTypeNative:record_daemon_version: 4850000
AppleMetaNodeLocation: /Local/Default
GeneratedUID: C1EAF7EF-850C-4166-A787-68FCCF32A093
GroupMembers: E75C1E93-9DD8-4E75-907A-EF5E09BED5C6 0E085ED3-F9C3-49BD-ABD1-F0B5579B11E4 B33564BB-AB9B-41A9-8605-7493280E951E
GroupMembership: nixbld1 nixbld2 nixbld3
PrimaryGroupID: 30000
RealName:
 Nix build group for nix-daemon
RecordName: nixbld
RecordType: dsRecTypeStandard:Groups
```

Build users informations:

```
$ for i in 1 2 3; dscl . -read "/Users/nixbld$i"; end
dsAttrTypeNative:_writers_passwd: nixbld1
dsAttrTypeNative:accountPolicyData:
 <?xml version="1.0" encoding="UTF-8"?>
<!DOCTYPE plist PUBLIC "-//Apple//DTD PLIST 1.0//EN" "http://www.apple.com/DTDs/PropertyList-1.0.dtd">
<plist version="1.0">
<dict>
	<key>creationTime</key>
	<real>1512821497.425267</real>
	<key>failedLoginCount</key>
	<integer>0</integer>
	<key>failedLoginTimestamp</key>
	<integer>0</integer>
	<key>passwordLastSetTime</key>
	<real>1512821498.062953</real>
</dict>
</plist>

dsAttrTypeNative:IsHidden: 1
dsAttrTypeNative:record_daemon_version: 4850000
AppleMetaNodeLocation: /Local/Default
GeneratedUID: E75C1E93-9DD8-4E75-907A-EF5E09BED5C6
NFSHomeDirectory: /var/empty
Password: ********
PrimaryGroupID: nixbld
RealName:
 Nix build user 1
RecordName: nixbld1
RecordType: dsRecTypeStandard:Users
UniqueID: 30001
UserShell: /sbin/nologin
dsAttrTypeNative:_writers_passwd: nixbld2
dsAttrTypeNative:accountPolicyData:
 <?xml version="1.0" encoding="UTF-8"?>
<!DOCTYPE plist PUBLIC "-//Apple//DTD PLIST 1.0//EN" "http://www.apple.com/DTDs/PropertyList-1.0.dtd">
<plist version="1.0">
<dict>
	<key>creationTime</key>
	<real>1512821735.5316291</real>
	<key>failedLoginCount</key>
	<integer>0</integer>
	<key>failedLoginTimestamp</key>
	<integer>0</integer>
	<key>passwordLastSetTime</key>
	<real>1512821736.1486039</real>
</dict>
</plist>

dsAttrTypeNative:IsHidden: 1
dsAttrTypeNative:record_daemon_version: 4850000
AppleMetaNodeLocation: /Local/Default
GeneratedUID: 0E085ED3-F9C3-49BD-ABD1-F0B5579B11E4
NFSHomeDirectory: /var/empty
Password: ********
PrimaryGroupID: nixbld
RealName:
 Nix build user 2
RecordName: nixbld2
RecordType: dsRecTypeStandard:Users
UniqueID: 30002
UserShell: /sbin/nologin
dsAttrTypeNative:_writers_passwd: nixbld3
dsAttrTypeNative:accountPolicyData:
 <?xml version="1.0" encoding="UTF-8"?>
<!DOCTYPE plist PUBLIC "-//Apple//DTD PLIST 1.0//EN" "http://www.apple.com/DTDs/PropertyList-1.0.dtd">
<plist version="1.0">
<dict>
	<key>creationTime</key>
	<real>1512821768.097223</real>
	<key>failedLoginCount</key>
	<integer>0</integer>
	<key>failedLoginTimestamp</key>
	<integer>0</integer>
	<key>passwordLastSetTime</key>
	<real>1512821768.806519</real>
</dict>
</plist>

dsAttrTypeNative:IsHidden: 1
dsAttrTypeNative:record_daemon_version: 4850000
AppleMetaNodeLocation: /Local/Default
GeneratedUID: B33564BB-AB9B-41A9-8605-7493280E951E
NFSHomeDirectory: /var/empty
Password: ********
PrimaryGroupID: nixbld
RealName:
 Nix build user 3
RecordName: nixbld3
RecordType: dsRecTypeStandard:Users
UniqueID: 30003
UserShell: /sbin/nologin
```

And what I've done when I created these group/users is:

```
$ sudo dseditgroup -o create -r "Nix build group for nix-daemon" -i "30000" "nixbld"

$ sudo dscl . -create /Users/nixbld1 UniqueID 30001
$ sudo dscl . -create /Users/nixbld1 IsHidden 1
$ sudo dscl . -create /Users/nixbld1 NFSHomeDirectory /var/empty
$ sudo dscl . -create /Users/nixbld1 RealName "Nix build user 1"
$ sudo dscl . -create /Users/nixbld1 UserShell /sbin/nologin
$ sudo dseditgroup -o edit -t user -a nixbld1 nixbld
$ sudo dscl . -create /Users/nixbld1 PrimaryGroupID 30000

# and for nixbld2 and nixbld3 (I have 4 cores but I only wants Nix uses most 3 cores, so I only create 3 build users)
```

And the group and users appears in System Preferences > Users and Groups, screenshot below:

![2017-12-09 21 00 55](https://user-images.githubusercontent.com/841969/33795852-2a9057bc-dd24-11e7-90f6-eea77f1d5f2e.png)

**Comment by david50407** [comment]
BTW, uses `id` command also report `no such user` for build users:

```
$ > for i in 1 2 3; id "nixbld$i"; end
id: nixbld1: no such user
id: nixbld2: no such user
id: nixbld3: no such user
```

**Comment by grahamc** [comment]
Did you follow both components of the user group membership?

if dseditgroup -o checkmember -m "$username" "$NIX_BUILD_GROUP_NAME" >
/dev/null 2>&1 ; then
row " Member of $NIX_BUILD_GROUP_NAME" "Yes"
else
_sudo "Add $username to the $NIX_BUILD_GROUP_NAME group"\
/usr/sbin/dseditgroup -o edit -t user \
-a "$username" "$NIX_BUILD_GROUP_NAME"
row " Member of $NIX_BUILD_GROUP_NAME" "Yes"
fi

if [ "$(dsclattr "$dsclpath" "PrimaryGroupID")" = "$NIX_BUILD_GROUP_ID" ];
then
row " PrimaryGroupID" "$NIX_BUILD_GROUP_ID"
else
_sudo "to let the nix daemon use this user for builds (this might seem
redundant, but there are two concepts of group membership)" \
/usr/bin/dscl . -create "$dsclpath" "PrimaryGroupID" "$NIX_BUILD_GROUP_ID"
row " PrimaryGroupID" "$NIX_BUILD_GROUP_ID"

fi
}

**Comment by david50407** [comment]
@grahamc 

yes, `nixbld1`, `nixbld2`, `nixbld3` are all in `nixbld` group: 

```
$ for i in 1 2 3; dseditgroup -o checkmember -m "nixbld$i" "nixbld"; end
yes nixbld1 is a member of nixbld
yes nixbld2 is a member of nixbld
yes nixbld3 is a member of nixbld

$ for i in 1 2 3; dscl . -read "/Users/nixbld$i" | grep PrimaryGroupID ; end
PrimaryGroupID: 30000
PrimaryGroupID: 30000
PrimaryGroupID: 30000
```

BUT, `dsclattr` command is not available in my system:

```
$ dsclattr
fish: Unknown command 'dsclattr'
```

**Comment by david50407** [comment]
oh wait.... after rebooting my Mac, it works. lol

**Comment by david50407** [comment]
Thank you @grahamc 

BUT `dsclattr` command seems not available in 10.13.2, maybe we need alternative command to check users informations?

**Comment by grahamc** [comment]
Ah, dsclattr is a function defined inside the install script :)
